### ISSUE TITLE
Plot the queuing delay and throughput while replaying the recorded traffic

### ISSUE THREAD

**Issue by spaul13** [ravinet/mahimahi#130]
Using the record shell, I have recorded the traffic for a webpage. But while replaying and plotting the queuing delay and throughput, mm-link requires an uplink and downlink file. 

But I have found any uplink and downlink file in my recorded traffic. Can anyone please tell me how to plot the recorded traffic using the replayshell?

**Comment by keithw** [comment]
Hello,

The `mm-webrecord` program records the resources from a Web download (HTTP and HTTPS URLs from a GET request, and the IP addresses that served them) -- it does not really record "traffic" per se.

You can recreate the same environment using the `mm-webreplay` command -- this will set up a bunch of Apache webservers in a container, with each webserver taking on the IP address of a real-life webserver from the recording, and serving the same resources at the same URLs.

If you want to record *traffic* over an emulated network, you can run the `mm-webreplay` command, and then inside that, run a sequence of link emulators -- e.g. the `mm-delay`, `mm-loss`, `mm-onoff` or `mm-link` commands. And then inside all of that, run the Web client and have it fetch a page from the webservers.

The `mm-link` command lets you specify a queue discipline and the delivery schedule for the uplink and downlink. (You can make a link with constant bandwidth if you want, or variable...) And it can display the results in real time (try the `mm-link --meter-all` option) or by logging to a file. If you log the uplink or downlinks to a file, you can plot them with the `mm-throughput-graph` or `mm-delay-graph` commands.

The bottom line is that you have to run over an emulated queue (as provided by the mm-link command) in order to be able to measure the queuing delay.

Hope this is helpful!

**Comment by spaul13** [comment]
@keithw thanks for the detailed reply. Its really helpful. Inside the replay shell I have tried mm-link --meter-all option but still showing invalid argument error. Here is the snippet

[replay] sj@p:~/mahimahi$ mm-link --meter-all
Usage: mm-link UPLINK-TRACE DOWNLINK-TRACE [OPTION]... [COMMAND]

Options = --once
          --uplink-log=FILENAME --downlink-log=FILENAME
          --meter-uplink --meter-uplink-delay
          --meter-downlink --meter-downlink-delay
          --meter-all
          --uplink-queue=QUEUE_TYPE --downlink-queue=QUEUE_TYPE
          --uplink-queue-args=QUEUE_ARGS --downlink-queue-args=QUEUE_ARGS

          QUEUE_TYPE = infinite | droptail | drophead | codel | pie
          QUEUE_ARGS = "NAME=NUMBER[, NAME2=NUMBER2, ...]"
              (with NAME = bytes | packets | target | interval | qdelay_ref | max_burst)
                  target, interval, qdelay_ref, max_burst are in milli-second

Died on std::runtime_error: invalid arguments

can u please help me telling where I am wrong. Inside the replay shell if I run ifconfig I can see all the contacted servers (virtually created apache servers).

One another query, how to log the HTTP uplink and downlink traffic differently into two different files using the record shell.

Any help will highly be appreciated. Thanks again....

**Comment by keithw** [comment]
Hello,

mm-link needs you to tell it what kind of link to emulate -- e.g., how fast can packets be delivered across the uplink and downlink, and what queue discipline to use in each direction.

We have some example uplink and downlink traces in the /usr/share/mahimahi/traces directory (which may be /usr/local/share/mahimahi/traces if you installed from source). You can also use a much simpler trace, e.g. if you want to allow 1 MTU-sized packet per millisecond (about 12 megabits per second), just use a trace file that has a single line with the number "1" on it. The mm-link(1) man page has a little more info.

**Comment by spaul13** [comment]
Thanks for your reply. I have tried the example traces and its working. But can you please tell me how can I generate my own uplink and downlink HTTP traffic traces using record shell.

**Comment by keithw** [comment]
Sorry, I'm not quite sure what you are asking for. The `mm-webrecord` program saves websites so they can be replayed later. (Basically it just records the IP addresses of webservers, and the response they provided in response to a particular GET request.) It doesn't record "traffic" per se.

**Comment by keithw** [comment]
The files that `mm-webrecord` generates are not uplink or downlink traces. These files each contain a particular HTTP request and its associated response. They are unrelated to the trace files used by `mm-link`.

**Comment by spaul13** [comment]
Then is it possible to plot the throughput and the queuing delay using these files in replay shell? As I found mm-link strictly requires uplink and downlink traces as I mentioned above also. Thanks again for this patient help.

**Comment by keithw** [comment]
I'm sorry, I still don't understand what you are asking for. The files written by mm-webrecord are just HTTP requests and responses -- they don't have a throughput or a queueing delay.

The mm-link program lets you emulate a limited-capacity link, of whatever capacity/bandwidth you want and with a particular queue discipline. You can run any program you want over this link, and it will capture the queueing delay of every packet (and the size and time of every packet, which you can use to compute throughput). It works with Web traffic (talking to a real webserver), or Web traffic (talking to mm-webreplay), or any other network program.

**Comment by spaul13** [comment]
my previous query got clarified. But can you please tell me, using mm-link how can I emulate 20 Mbps link rate with codel queue type and store the captured queuing delay and throughput (downlink) in a file. It will immensely help my understanding. Thanks a lot again.

**Comment by cbothra123** [comment]
How can we emulate 5 Mbps if the MTU size is 1500 bytes?
